tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -0,0 +1,183 @@
+    theBuilder.decorateLocation(input.first, nextInputLocation++);
+  }
+  for (const auto &output : stageOutputs) {
+    theBuilder.decorateLocation(output.first, nextOutputLocation++);
+  }
+}
+
+QualType
+DeclResultIdMapper::getFnParamOrRetType(const DeclaratorDecl *decl) const {
+  if (const auto *funcDecl = dyn_cast<FunctionDecl>(decl)) {
+    return funcDecl->getReturnType();
+  }
+  return decl->getType();
+}
+
+void DeclResultIdMapper::createStageVariables(const DeclaratorDecl *decl,
+                                              bool actAsInput) {
+  QualType type = getFnParamOrRetType(decl);
+
+  if (type->isVoidType()) {
+    // No stage variables will be created for void type.
+    return;
+  }
+
+  const std::string semantic = getStageVarSemantic(decl);
+  if (!semantic.empty()) {
+    // Found semantic attached directly to this Decl. This means we need to
+    // map this decl to a single stage variable.
+    const uint32_t typeId = typeTranslator.translateType(type);
+    const auto kind = getStageVarKind(semantic);
+
+    if (actAsInput) {
+      // Stage (builtin) input variable cases
+      const uint32_t varId =
+          theBuilder.addStageIOVariable(typeId, spv::StorageClass::Input);
+
+      stageInputs.push_back(std::make_pair(varId, semantic));
+      remappedDecls[decl] = varId;
+      stageVars.insert(varId);
+    } else {
+      // Handle output builtin variables first
+      if (shaderStage == spv::ExecutionModel::Vertex &&
+          kind == StageVarKind::Position) {
+        const uint32_t varId =
+            theBuilder.addStageBuiltinVariable(typeId, spv::BuiltIn::Position);
+
+        stageBuiltins.push_back(std::make_pair(varId, semantic));
+        remappedDecls[decl] = varId;
+        stageVars.insert(varId);
+      } else {
+        // The rest are normal stage output variables
+        const uint32_t varId =
+            theBuilder.addStageIOVariable(typeId, spv::StorageClass::Output);
+
+        stageOutputs.push_back(std::make_pair(varId, semantic));
+        remappedDecls[decl] = varId;
+        stageVars.insert(varId);
+      }
+    }
+  } else {
+    // If the decl itself doesn't have semantic, it should be a struct having
+    // all its fields with semantics.
+    assert(type->isStructureType() &&
+           "found non-struct decls without semantics");
+
+    const auto *structDecl = cast<RecordType>(type.getTypePtr())->getDecl();
+
+    // Recursively handle all the fields.
+    for (const auto *field : structDecl->fields()) {
+      createStageVariables(field, actAsInput);
+    }
+  }
+}
+
+DeclResultIdMapper::StageVarKind
+DeclResultIdMapper::getStageVarKind(llvm::StringRef semantic) const {
+  return llvm::StringSwitch<StageVarKind>(semantic)
+      .Case("", StageVarKind::None)
+      .StartsWith("COLOR", StageVarKind::Color)
+      .StartsWith("POSITION", StageVarKind::Position)
+      .StartsWith("SV_POSITION", StageVarKind::Position)
+      .StartsWith("SV_TARGET", StageVarKind::Target)
```

Comment:
Yep. IMHO, it's not really necessary. It will be driven by implementing other features and we'll surely do it. I typically view `TODO`s as a notice that the current impl. is kinda sub-optimal or there exist corner cases yet to be covered. It's basically a reminder thing.